File: apps/web/src/lib/scenario-builder.ts
import type { ConfidenceLevel, PlanResponse, PlanSlice } from "../types/plan";
import type {
  ScenarioRecommendation,
  ScenarioRecommendationAdjustment,
} from "../types/scenario";

export interface ScenarioChannelBaseline {
  channel: string;
  spend: number;
  revenue: number;
  confidence: ConfidenceLevel;
  baseRoi: number | null;
}

export interface ScenarioBaseline {
  channels: ScenarioChannelBaseline[];
  totalSpend: number;
  totalRevenue: number;
  horizonDays: number;
}

export interface ScenarioAdjustmentMap {
  [channel: string]: number;
}

export interface ScenarioChannelOutcome {
  channel: string;
  confidence: ConfidenceLevel;
  baseSpend: number;
  baseRevenue: number;
  scenarioSpend: number;
  scenarioRevenue: number;
  deltaSpend: number;
  deltaRevenue: number;
  baseRoi: number | null;
  scenarioRoi: number | null;
}

export interface ScenarioOutcomeSummary {
  totalBaseSpend: number;
  totalScenarioSpend: number;
  totalBaseRevenue: number;
  totalScenarioRevenue: number;
  deltaSpend: number;
  deltaRevenue: number;
  baseRoi: number | null;
  scenarioRoi: number | null;
  weightedConfidence: ConfidenceLevel;
}

export interface ScenarioOutcome {
  summary: ScenarioOutcomeSummary;
  channels: ScenarioChannelOutcome[];
}

const CONFIDENCE_ORDER: ConfidenceLevel[] = ["LOW", "MEDIUM", "HIGH"];
const CONFIDENCE_WEIGHT: Record<ConfidenceLevel, number> = {
  HIGH: 3,
  MEDIUM: 2,
  LOW: 1,
};
const CONFIDENCE_SCORE: Record<ConfidenceLevel, number> = {
  HIGH: 1,
  MEDIUM: 0.6,
  LOW: 0.3,
};
const CONFIDENCE_RISK_ADJUSTMENT: Record<ConfidenceLevel, number> = {
  HIGH: 0.05,
  MEDIUM: 0.15,
  LOW: 0.3,
};

const DEFAULT_CONFIDENCE: ConfidenceLevel = "MEDIUM";

function normaliseConfidence(value: ConfidenceLevel | null | undefined): ConfidenceLevel {
  if (!value) {
    return DEFAULT_CONFIDENCE;
  }
  if (value === "HIGH" || value === "MEDIUM" || value === "LOW") {
    return value;
  }
  return DEFAULT_CONFIDENCE;
}

function sumSlices<T extends PlanSlice>(
  slices: T[],
  selector: (slice: T) => number | null | undefined,
): number {
  return slices.reduce((total, slice) => {
    const value = selector(slice);
    const numeric = typeof value === "number" && Number.isFinite(value) ? value : 0;
    return total + numeric;
  }, 0);
}

function computeChannelConfidence(slices: PlanSlice[]): ConfidenceLevel {
  if (slices.length === 0) {
    return DEFAULT_CONFIDENCE;
  }
  const scores = slices.map((slice) => CONFIDENCE_WEIGHT[normaliseConfidence(slice.confidence)]);
  const worstScore = Math.min(...scores);
  return CONFIDENCE_ORDER.find((level) => CONFIDENCE_WEIGHT[level] === worstScore) ?? DEFAULT_CONFIDENCE;
}

function computeRoi(spend: number, revenue: number): number | null {
  if (!Number.isFinite(spend) || spend <= 0) {
    return null;
  }
  if (!Number.isFinite(revenue)) {
    return null;
  }
  return revenue / spend;
}

export function buildScenarioBaseline(plan: PlanResponse): ScenarioBaseline {
  const channelGroups = new Map<string, PlanSlice[]>();

  for (const slice of plan.slices) {
    const channel = slice.channel?.trim() || "Unassigned channel";
    const existing = channelGroups.get(channel);
    if (existing) {
      existing.push(slice);
    } else {
      channelGroups.set(channel, [slice]);
    }
  }

  const channels: ScenarioChannelBaseline[] = Array.from(channelGroups.entries()).map(([channel, slices]) => {
    const spend = sumSlices(slices, (slice) => slice.recommended_spend);
    const revenue = sumSlices(slices, (slice) => slice.expected_revenue?.p50 ?? null);
    const confidence = computeChannelConfidence(slices);
    return {
      channel,
      spend,
      revenue,
      confidence,
      baseRoi: computeRoi(spend, revenue),
    };
  });

  channels.sort((a, b) => b.spend - a.spend);

  const totalSpend = channels.reduce((total, channel) => total + channel.spend, 0);
  const totalRevenue = channels.reduce((total, channel) => total + channel.revenue, 0);

  return {
    channels,
    totalSpend,
    totalRevenue,
    horizonDays: plan.horizon_days,
  };
}

function clampMultiplier(value: number): number {
  if (!Number.isFinite(value)) {
    return 1;
  }
  return Math.max(0, value);
}

function deriveWeightedConfidence(
  channels: ScenarioChannelOutcome[],
  totalScenarioSpend: number,
): ConfidenceLevel {
  if (!Number.isFinite(totalScenarioSpend) || totalScenarioSpend <= 0) {
    return DEFAULT_CONFIDENCE;
  }

  const weightedScore = channels.reduce((total, channel) => {
    const weight = channel.scenarioSpend / totalScenarioSpend;
    return total + weight * CONFIDENCE_WEIGHT[channel.confidence];
  }, 0);

  if (weightedScore >= 2.5) {
    return "HIGH";
  }
  if (weightedScore >= 1.75) {
    return "MEDIUM";
  }
  return "LOW";
}

export function applyScenarioAdjustments(
  baseline: ScenarioBaseline,
  adjustments: ScenarioAdjustmentMap,
): ScenarioOutcome {
  const channels: ScenarioChannelOutcome[] = baseline.channels.map((channel) => {
    const multiplier = clampMultiplier(adjustments[channel.channel] ?? 1);
    const delta = multiplier - 1;
    const risk = CONFIDENCE_RISK_ADJUSTMENT[channel.confidence] ?? CONFIDENCE_RISK_ADJUSTMENT.MEDIUM;

    let revenueMultiplier: number;
    if (delta >= 0) {
      revenueMultiplier = 1 + delta * (1 - risk);
    } else {
      revenueMultiplier = 1 + delta * (1 + risk);
    }
    if (!Number.isFinite(revenueMultiplier) || revenueMultiplier < 0) {
      revenueMultiplier = 0;
    }

    const scenarioSpend = channel.spend * multiplier;
    const scenarioRevenue = channel.revenue * revenueMultiplier;
    const baseSpend = channel.spend;
    const baseRevenue = channel.revenue;

    return {
      channel: channel.channel,
      confidence: channel.confidence,
      baseSpend,
      baseRevenue,
      scenarioSpend,
      scenarioRevenue,
      deltaSpend: scenarioSpend - baseSpend,
      deltaRevenue: scenarioRevenue - baseRevenue,
      baseRoi: channel.baseRoi,
      scenarioRoi: computeRoi(scenarioSpend, scenarioRevenue),
    };
  });

  const totalBaseSpend = baseline.totalSpend;
  const totalBaseRevenue = baseline.totalRevenue;
  const totalScenarioSpend = channels.reduce((total, channel) => total + channel.scenarioSpend, 0);
  const totalScenarioRevenue = channels.reduce((total, channel) => total + channel.scenarioRevenue, 0);

  const summary: ScenarioOutcomeSummary = {
    totalBaseSpend,
    totalScenarioSpend,
    totalBaseRevenue,
    totalScenarioRevenue,
    deltaSpend: totalScenarioSpend - totalBaseSpend,
    deltaRevenue: totalScenarioRevenue - totalBaseRevenue,
    baseRoi: computeRoi(totalBaseSpend, totalBaseRevenue),
    scenarioRoi: computeRoi(totalScenarioSpend, totalScenarioRevenue),
    weightedConfidence: deriveWeightedConfidence(channels, totalScenarioSpend),
  };

  return {
    summary,
    channels,
  };
}

function describeConfidence(level: ConfidenceLevel): string {
  switch (level) {
    case "HIGH":
      return "high";
    case "MEDIUM":
      return "moderate";
    default:
      return "low";
  }
}

function deriveChannelRoi(channel: ScenarioChannelBaseline): number {
  if (channel.baseRoi && Number.isFinite(channel.baseRoi)) {
    return channel.baseRoi;
  }
  if (channel.spend <= 0) {
    return 0;
  }
  return channel.revenue / channel.spend;
}

function buildAdjustmentsFromChannels(
  channels: ScenarioChannelBaseline[],
  options: { increase: boolean; bump: number },
): ScenarioRecommendationAdjustment[] {
  const { increase, bump } = options;
  if (!Number.isFinite(bump) || bump <= 0) {
    return [];
  }
  return channels.map((channel) => {
    const roi = deriveChannelRoi(channel);
    const multiplier = increase ? 1 + bump : 1 - bump;
    const direction = increase ? "Increase" : "Reduce";
    const formattedRoi = Number.isFinite(roi) && roi > 0 ? `${roi.toFixed(1)}x ROI` : "ROI signal building";
    return {
      channel: channel.channel,
      multiplier,
      confidence: channel.confidence,
      rationale: `${direction} by ${Math.round(bump * 100)}% (${formattedRoi}, ${describeConfidence(channel.confidence)} confidence).`,
    };
  });
}

export function deriveScenarioRecommendations(baseline: ScenarioBaseline): ScenarioRecommendation[] {
  if (!baseline.channels.length) {
    return [];
  }

  const roiValues = baseline.channels
    .map((channel) => deriveChannelRoi(channel))
    .filter((value) => Number.isFinite(value) && value > 0);
  const averageRoi =
    roiValues.length > 0
      ? roiValues.reduce((total, value) => total + value, 0) / roiValues.length
      : 0;

  const confidenceValues = baseline.channels.map((channel) => CONFIDENCE_SCORE[channel.confidence]);
  const averageConfidence =
    confidenceValues.reduce((total, value) => total + value, 0) / confidenceValues.length;

  const growthCandidates = baseline.channels.filter((channel) => {
    const roi = deriveChannelRoi(channel);
    const confidenceScore = CONFIDENCE_SCORE[channel.confidence];
    return roi >= averageRoi && confidenceScore >= Math.max(averageConfidence, 0.55);
  });

  const sortedByRoi = [...baseline.channels].sort(
    (a, b) => deriveChannelRoi(b) - deriveChannelRoi(a),
  );

  if (growthCandidates.length === 0 && sortedByRoi.length > 0) {
    growthCandidates.push(sortedByRoi[0]);
  }

  const trimCandidates = baseline.channels.filter((channel) => {
    const roi = deriveChannelRoi(channel);
    const confidenceScore = CONFIDENCE_SCORE[channel.confidence];
    return confidenceScore <= Math.max(averageConfidence - 0.05, 0.45) || roi <= averageRoi * 0.85;
  });

  if (trimCandidates.length === 0 && sortedByRoi.length > 0) {
    trimCandidates.push(sortedByRoi[sortedByRoi.length - 1]);
  }

  let rebalancePair: ScenarioChannelBaseline[] = [];
  if (sortedByRoi.length >= 2) {
    const top = sortedByRoi[0];
    const bottom = sortedByRoi[sortedByRoi.length - 1];
    if (top.channel !== bottom.channel) {
      rebalancePair = [top, bottom];
    }
  }

  const recommendations: ScenarioRecommendation[] = [
    {
      id: "accelerate_high_confidence",
      label: "Accelerate high-confidence reach",
      description: "Increase exposure on channels with resilient ROI and strong weather confidence.",
      adjustments: buildAdjustmentsFromChannels(growthCandidates, { increase: true, bump: 0.15 }),
      tags: ["growth", "high-confidence"],
    },
    {
      id: "stabilise_low_signal",
      label: "Stabilise low-signal spend",
      description: "Dial back budget from channels showing weak ROI or low confidence until telemetry firms up.",
      adjustments: buildAdjustmentsFromChannels(trimCandidates, { increase: false, bump: 0.1 }),
      tags: ["risk", "confidence"],
    },
  ];

  if (rebalancePair.length === 2) {
    const [increaseChannel, decreaseChannel] = rebalancePair;
    recommendations.push({
      id: "rebalance_mix",
      label: "Rebalance mix for weather swings",
      description: "Shift budget toward resilient performers while easing pressure on softer channels.",
      adjustments: [
        ...buildAdjustmentsFromChannels([increaseChannel], { increase: true, bump: 0.12 }),
        ...buildAdjustmentsFromChannels([decreaseChannel], { increase: false, bump: 0.08 }),
      ],
      tags: ["balanced", "mix-shift"],
    });
  }

  return recommendations;
}

File: shared/services/scenario_builder.py
"""
Scenario analysis utilities shared between API and worker layers.

These helpers mirror the scenario-builder logic used in the web client so
back-end services can generate consistent summaries when persisting or
evaluating what-if mixes.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Mapping

from shared.schemas.base import ConfidenceLevel, PlanResponse, PlanSlice


# Confidence scoring mirrors the TypeScript implementation to keep outputs
# consistent across front-end and back-end callers.
_CONFIDENCE_WEIGHT: dict[ConfidenceLevel, int] = {
    ConfidenceLevel.HIGH: 3,
    ConfidenceLevel.MEDIUM: 2,
    ConfidenceLevel.LOW: 1,
}

_CONFIDENCE_RISK_ADJUSTMENT: dict[ConfidenceLevel, float] = {
    ConfidenceLevel.HIGH: 0.05,
    ConfidenceLevel.MEDIUM: 0.15,
    ConfidenceLevel.LOW: 0.3,
}


@dataclass(slots=True)
class ScenarioChannelBaseline:
    channel: str
    spend: float
    revenue: float
    confidence: ConfidenceLevel
    base_roi: float | None


@dataclass(slots=True)
class ScenarioBaseline:
    channels: list[ScenarioChannelBaseline]
    total_spend: float
    total_revenue: float
    horizon_days: int


@dataclass(slots=True)
class ScenarioChannelOutcome:
    channel: str
    confidence: ConfidenceLevel
    base_spend: float
    base_revenue: float
    scenario_spend: float
    scenario_revenue: float
    delta_spend: float
    delta_revenue: float
    base_roi: float | None
    scenario_roi: float | None


@dataclass(slots=True)
class ScenarioOutcomeSummary:
    total_base_spend: float
    total_scenario_spend: float
    total_base_revenue: float
    total_scenario_revenue: float
    delta_spend: float
    delta_revenue: float
    base_roi: float | None
    scenario_roi: float | None
    weighted_confidence: ConfidenceLevel


@dataclass(slots=True)
class ScenarioOutcome:
    summary: ScenarioOutcomeSummary
    channels: list[ScenarioChannelOutcome]


def _normalise_confidence(value: ConfidenceLevel | None) -> ConfidenceLevel:
    if value is None:
        return ConfidenceLevel.MEDIUM
    if isinstance(value, ConfidenceLevel):
        return value
    # ConfidenceLevel is an enum, but callers might pass str values when
    # validated via pydantic; coerce cautiously.
    try:
        return ConfidenceLevel(value)
    except ValueError:  # pragma: no cover - defensive branch
        return ConfidenceLevel.MEDIUM


def _compute_roi(spend: float, revenue: float) -> float | None:
    if spend <= 0 or not float(spend):
        return None
    if not float(revenue):
        return None
    return revenue / spend


def _sum_slices(slices: Iterable[PlanSlice], attr: str) -> float:
    total = 0.0
    for slice_ in slices:
        value = getattr(slice_, attr)
        if isinstance(value, (int, float)):
            total += float(value)
        elif value is not None:
            # expected_revenue / expected_roas are PlanQuantiles with p10/p50/p90
            p50 = getattr(value, "p50", None)
            if isinstance(p50, (int, float)):
                total += float(p50)
    return total


def _compute_channel_confidence(slices: Iterable[PlanSlice]) -> ConfidenceLevel:
    worst_weight = None
    for slice_ in slices:
        weight = _CONFIDENCE_WEIGHT[_normalise_confidence(slice_.confidence)]
        if worst_weight is None or weight < worst_weight:
            worst_weight = weight
    if worst_weight is None:
        return ConfidenceLevel.MEDIUM
    for level, weight in _CONFIDENCE_WEIGHT.items():
        if weight == worst_weight:
            return level
    return ConfidenceLevel.MEDIUM


def build_scenario_baseline(plan: PlanResponse) -> ScenarioBaseline:
    """Aggregate plan slices into per-channel baselines."""

    channels: list[ScenarioChannelBaseline] = []
    slices_by_channel: dict[str, list[PlanSlice]] = {}

    for slice_ in plan.slices:
        channel = (slice_.channel or "Unassigned channel").strip()
        slices_by_channel.setdefault(channel, []).append(slice_)

    for channel, slices in slices_by_channel.items():
        spend = _sum_slices(slices, "recommended_spend")
        revenue = _sum_slices(slices, "expected_revenue")
        confidence = _compute_channel_confidence(slices)
        channels.append(
            ScenarioChannelBaseline(
                channel=channel,
                spend=spend,
                revenue=revenue,
                confidence=confidence,
                base_roi=_compute_roi(spend, revenue),
            )
        )

    channels.sort(key=lambda item: item.spend, reverse=True)

    total_spend = sum(channel.spend for channel in channels)
    total_revenue = sum(channel.revenue for channel in channels)

    return ScenarioBaseline(
        channels=channels,
        total_spend=total_spend,
        total_revenue=total_revenue,
        horizon_days=plan.horizon_days,
    )


def _clamp_multiplier(value: float) -> float:
    if not isinstance(value, (int, float)):
        return 1.0
    if value < 0:
        return 0.0
    return float(value)


def _derive_weighted_confidence(
    channels: Iterable[ScenarioChannelOutcome],
    total_scenario_spend: float,
) -> ConfidenceLevel:
    if total_scenario_spend <= 0:
        return ConfidenceLevel.MEDIUM

    weighted_score = 0.0
    for channel in channels:
        weight = channel.scenario_spend / total_scenario_spend
        weighted_score += weight * _CONFIDENCE_WEIGHT[channel.confidence]

    if weighted_score >= 2.5:
        return ConfidenceLevel.HIGH
    if weighted_score >= 1.75:
        return ConfidenceLevel.MEDIUM
    return ConfidenceLevel.LOW


def apply_scenario_adjustments(
    baseline: ScenarioBaseline,
    adjustments: Mapping[str, float],
) -> ScenarioOutcome:
    """Apply per-channel multipliers to the baseline and return outcome deltas."""

    channel_outcomes: list[ScenarioChannelOutcome] = []

    for channel in baseline.channels:
        multiplier = _clamp_multiplier(adjustments.get(channel.channel, 1.0))
        delta = multiplier - 1.0
        risk = _CONFIDENCE_RISK_ADJUSTMENT[channel.confidence]

        if delta >= 0:
            revenue_multiplier = 1.0 + delta * (1.0 - risk)
        else:
            revenue_multiplier = 1.0 + delta * (1.0 + risk)

        revenue_multiplier = max(revenue_multiplier, 0.0)

        scenario_spend = channel.spend * multiplier
        scenario_revenue = channel.revenue * revenue_multiplier

        channel_outcomes.append(
            ScenarioChannelOutcome(
                channel=channel.channel,
                confidence=channel.confidence,
                base_spend=channel.spend,
                base_revenue=channel.revenue,
                scenario_spend=scenario_spend,
                scenario_revenue=scenario_revenue,
                delta_spend=scenario_spend - channel.spend,
                delta_revenue=scenario_revenue - channel.revenue,
                base_roi=channel.base_roi,
                scenario_roi=_compute_roi(scenario_spend, scenario_revenue),
            )
        )

    total_base_spend = baseline.total_spend
    total_base_revenue = baseline.total_revenue
    total_scenario_spend = sum(channel.scenario_spend for channel in channel_outcomes)
    total_scenario_revenue = sum(channel.scenario_revenue for channel in channel_outcomes)

    summary = ScenarioOutcomeSummary(
        total_base_spend=total_base_spend,
        total_scenario_spend=total_scenario_spend,
        total_base_revenue=total_base_revenue,
        total_scenario_revenue=total_scenario_revenue,
        delta_spend=total_scenario_spend - total_base_spend,
        delta_revenue=total_scenario_revenue - total_base_revenue,
        base_roi=_compute_roi(total_base_spend, total_base_revenue),
        scenario_roi=_compute_roi(total_scenario_spend, total_scenario_revenue),
        weighted_confidence=_derive_weighted_confidence(
            channel_outcomes, total_scenario_spend
        ),
    )

    return ScenarioOutcome(summary=summary, channels=channel_outcomes)


__all__ = [
    "ScenarioBaseline",
    "ScenarioChannelBaseline",
    "ScenarioChannelOutcome",
    "ScenarioOutcome",
    "ScenarioOutcomeSummary",
    "apply_scenario_adjustments",
    "build_scenario_baseline",
]
